Remove design time helpers from C# completion list

allisonchou opened the pull request:
### Summary of the changes
 - Removes Razor design time helpers from C#'s completion list since they will not be applicable.
 - If the current identifier being targeted does not start with a double underscore, we trim out all items starting with "__" from the completion list. If the current identifier does start with a double underscore (e.g. "__ab[||]"), we only trim out common design time helpers from the completion list.
 
Definitely open to adding more items to the "common" design time helpers list. I mainly just used the recommended ones from the original issue.

Fixes: 
https://github.com/dotnet/aspnetcore/issues/25091

allisonchou commented on src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/HtmlCSharp/CompletionHandler.cs:
```diff
@@ -153,27 +148,25 @@ internal class CompletionHandler : IRequestHandler<CompletionParams, SumType<Com
         }
 
-        private bool IsSimpleImplicitExpression(LSPDocumentSnapshot documentSnapshot, CompletionParams request, LanguageServerKind serverKind)
+        private TextExtent? GetWordExtent(LSPDocumentSnapshot documentSnapshot, CompletionParams request, LanguageServerKind serverKind)
```
I extracted the word extent logic out from `IsSimpleImplicitExpression` since the same logic also applies to removing design time items.

NTaylorMullen replied:
This actually looks like a really good candidate for a general ITextSnapshot extension if you want to refactor it to something there.

I'd change the signature though to be more generic. Something like:

```C#
private static TextExtent? GetWordExtent(this ITextSnapshot textSnapshot, int character, int line, ITextStructureNavigatorSelectorService textStructureNavigatorService)
```

NTaylorMullen commented on src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/HtmlCSharp/CompletionHandler.cs:
```diff
@@ -153,27 +148,25 @@ internal class CompletionHandler : IRequestHandler<CompletionParams, SumType<Com
-        private bool IsSimpleImplicitExpression(LSPDocumentSnapshot documentSnapshot, CompletionParams request, LanguageServerKind serverKind)
+        private TextExtent? GetWordExtent(LSPDocumentSnapshot documentSnapshot, CompletionParams request, LanguageServerKind serverKind)
         {
             if (serverKind != LanguageServerKind.CSharp)
```
It looks like we're doing this serverkind C# check in several locations now. Lets wrap the IncludeCSharpKeywords bits and the RemoveDesignTimeItems bit in the check, that way each method doesn't need to recheck the same thing. Could even refactor all the C# massaging logic into its own method since it's conceptually a different stage of completion

NTaylorMullen commented on src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/HtmlCSharp/CompletionHandler.cs:
```diff
@@ -215,6 +229,63 @@ private bool IsSimpleImplicitExpression(LSPDocumentSnapshot documentSnapshot, Co
+                if (wordExtent.HasValue)
+                {
+                    var wordCharacterStartIndex = wordExtent.Value.Span.Start.Position;
+                    if (wordCharacterStartIndex + 1 < documentSnapshot.Snapshot.Length)
```
Hmm, feels like an odd way to check if the word is at least 2 characters. Probably makes more sense to do
```C#
var wordSpan = wordExtent.Value.Span;
if (wordSpan.Length >= 2)
{
...
}
```

NTaylorMullen commented on src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/HtmlCSharp/CompletionHandler.cs:
```diff
@@ -215,6 +229,63 @@ private bool IsSimpleImplicitExpression(LSPDocumentSnapshot documentSnapshot, Co
+                        // If the current identifier starts with "__", only trim out common design time helpers from the list.
+                        if (snapshot[wordCharacterStartIndex] == '_' && snapshot[wordCharacterStartIndex + 1] == '_')
+                        {
+                            return items.Except(designTimeHelpersCompletionItems, CompletionItemComparer.Instance).ToArray();
```
Store the result of this in a variable before returning to ease in debugging.

NTaylorMullen commented on src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/HtmlCSharp/CompletionHandler.cs:
```diff
@@ -215,6 +229,63 @@ private bool IsSimpleImplicitExpression(LSPDocumentSnapshot documentSnapshot, Co
+                }
+
+                // In all other cases, trim out both common design time helpers and all completion items starting with "__".
+                var updatedItems = items.Except(designTimeHelpersCompletionItems, CompletionItemComparer.Instance).ToArray();
```
```suggestion
                var filteredItems = items.Except(designTimeHelpersCompletionItems, CompletionItemComparer.Instance).ToArray();
```

NTaylorMullen replied:
No matter what you do in this flow we're always going to filter out the common design time helpers so it may be valuable and more readable to do that at the top and then simplify the flow to something like:
```C#
var filteredItems = items.Except(razorDesignTimeHelpers, CompletionItemComparer.Instance).ToArray();
if (SuppressAllDesignTimeItems(documentSnapshot.Snapshot, wordExtent))
{
    var designTimeSuppressedItems = filteredItems.Where(item => item.InsertText != null && !item.InsertText.StartsWith("__", StringComparison.Ordinal)).ToArray();
    return designTimeSupprssedItems;
}

return filteredItems;
```

NTaylorMullen commented on src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/HtmlCSharp/CompletionHandler.cs:
```diff
@@ -215,6 +229,63 @@ private bool IsSimpleImplicitExpression(LSPDocumentSnapshot documentSnapshot, Co
+
+            static CompletionItem[] RemoveDesignTimeItems(CompletionItem[] items, LSPDocumentSnapshot documentSnapshot, TextExtent? wordExtent)
+            {
+                var designTimeHelpersCompletionItems = GenerateCompletionItems(DesignTimeHelpers);
```
Any reason we can't generate this completion list once instead of per-request?

NTaylorMullen commented on src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/HtmlCSharp/CompletionHandler.cs:
```diff
@@ -215,6 +229,63 @@ private bool IsSimpleImplicitExpression(LSPDocumentSnapshot documentSnapshot, Co
+            => completionItems.Select(k => new CompletionItem
+            {
+                Label = k,
+                InsertText = k,
```
No need to provide InsertText, filterText or SortText, they default to Label according to the spec. Also don't need to provide the InserTextFormat

NTaylorMullen commented on src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/HtmlCSharp/CompletionHandler.cs:
```diff
@@ -35,15 +35,10 @@ internal class CompletionHandler : IRequestHandler<CompletionParams, SumType<Com
-            SortText = k,
-            InsertTextFormat = InsertTextFormat.Plaintext,
-        }).ToArray();
+            "__builder", "__o", "__RazorDirectiveTokenHelpers__", "BuildRenderTree"
```
We should probably include a few of the [non-Blazor directive helpers](https://github.com/dotnet/aspnetcore/blob/2af2bc2fa9c46bc8ba32a5555e3310bee43f4a70/src/Razor/Microsoft.AspNetCore.Razor.Language/test/TestFiles/IntegrationTests/CodeGenerationIntegrationTest/AttributeTargetingTagHelpers_DesignTime.codegen.cs) that aren't dynamic too:
- `__typeHelper`
- `__tagHelperExecutionContext`
- `__tagHelperRunner`

NTaylorMullen reviewed (changes requested):
Super excited to see this play out. Will improve the overall feeling of C# completion in typing scenarios and will drastically reduce the potential for the compiler to unintentionally introduce bugs

allisonchou commented on src/Razor/test/Microsoft.VisualStudio.LanguageServerClient.Razor.Test/HtmlCSharp/CompletionHandlerTest.cs:
```diff
@@ -233,7 +233,7 @@ public async Task HandleRequestAsync_CSharpProjection_DoNotPreselectAfterAt()
             // Arrange
             var called = false;
-            var expectedItem = new CompletionItem() { InsertText = "__builder", Label = "__builder", Preselect = true };
+            var expectedItem = new CompletionItem() { InsertText = "AccessViolationException", Label = "AccessViolationException", Preselect = true };
```
Technically `_Imports` is the first item in the list now that `__builder` is removed, but I changed this to `AccessViolationException` since `_Imports` technically shouldn't be in the list either (https://github.com/dotnet/aspnetcore/issues/29562).

allisonchou commented:
Thanks for the review Taylor! I've addressed your feedback (+ fixed up some failing tests), so the PR should be ready for another round of review whenever you have time.

NTaylorMullen commented on src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/HtmlCSharp/CompletionHandler.cs:
```diff
@@ -35,15 +33,15 @@ internal class CompletionHandler : IRequestHandler<CompletionParams, SumType<Com
-            SortText = k,
-            InsertTextFormat = InsertTextFormat.Plaintext,
-        }).ToArray();
+            "__builder", "__o", "__RazorDirectiveTokenHelpers__",
```
nit: To ease in reading I'd just separate these all by a newline

NTaylorMullen commented on src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/HtmlCSharp/CompletionHandler.cs:
```diff
@@ -153,59 +151,59 @@ internal class CompletionHandler : IRequestHandler<CompletionParams, SumType<Com
+        private SumType<CompletionItem[], CompletionList>? MassageCSharpCompletions(
+            CompletionParams request,
+            LSPDocumentSnapshot documentSnapshot,
+            SumType<CompletionItem[], CompletionList>? result)
```
This method depends on this not being `null` (otherwise .Value below would cause all sorts of chaos). I'd recommend making this non-nullable

NTaylorMullen commented on src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/HtmlCSharp/CompletionHandler.cs:
```diff
@@ -153,59 +151,59 @@ internal class CompletionHandler : IRequestHandler<CompletionParams, SumType<Com
         }
 
-        private bool IsSimpleImplicitExpression(LSPDocumentSnapshot documentSnapshot, CompletionParams request, LanguageServerKind serverKind)
+        private SumType<CompletionItem[], CompletionList>? MassageCSharpCompletions(
```
nit: the method signature of this is inconsistent with the other signatures. Some take DocumentSnapshot first others take it second. Normalizing it either way so it's consistent will make future reading/fixing easier.

NTaylorMullen commented on src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/HtmlCSharp/CompletionHandler.cs:
```diff
@@ -215,6 +213,67 @@ private bool IsSimpleImplicitExpression(LSPDocumentSnapshot documentSnapshot, Co
+                    }
+
+                    var snapshot = documentSnapshot.Snapshot;
+                    var startIndex = wordExtent.Value.Span.Start.Position;
```
```suggestion
                    var startIndex = wordSpan.Start.Position;
```

NTaylorMullen commented on src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/HtmlCSharp/CompletionHandler.cs:
```diff
@@ -215,6 +213,67 @@ private bool IsSimpleImplicitExpression(LSPDocumentSnapshot documentSnapshot, Co
+
+            return result;
+
+            static CompletionItem[] RemoveDesignTimeItems(CompletionItem[] items, LSPDocumentSnapshot documentSnapshot, TextExtent? wordExtent)
```
lol it's interesting having the nested nested local functions but I don't have a better recommendation. Overall love how much clearer the flow is.

NTaylorMullen commented on src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/ITextSnapshotExtensions.cs:
```diff
@@ -20,5 +22,33 @@ public static void GetLineAndCharacter(this ITextSnapshot snapshot, int index, o
         }
+
+        public static TextExtent? GetWordExtent(
+            this ITextSnapshot snapshot,
```
null check the snapshot at the top of this method, ditto with the `textStructureNavigatorService`

NTaylorMullen commented on src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/ITextSnapshotExtensions.cs:
```diff
@@ -20,5 +22,33 @@ public static void GetLineAndCharacter(this ITextSnapshot snapshot, int index, o
+
+        public static TextExtent? GetWordExtent(
+            this ITextSnapshot snapshot,
+            int line, int character,
```
nit: We typically separate all parameters by a newline even if they're somewhat related for consistencies sake

NTaylorMullen reviewed (approved):
👏 looks awesome! Only minor comments.
